test/lib/utils/joinHelper.ts

```diff
@@ -0,0 +1,401 @@
+        slippage,
+    } = txInput;
+
+    const { joinQueryResult, joinBuildOutput } = await sdkJoin({
+        poolJoin,
+        joinInput,
+        poolStateInput,
+        slippage,
+        testAddress,
+    });
+
+    // get tokens for balance change - pool tokens, BPT, native
+    const tokens = getTokens(poolStateInput);
+
+    // send transaction and calculate balance changes
+    const txOutput = await sendTransactionGetBalances(
+        tokens,
+        client,
+        testAddress,
+        joinBuildOutput.to,
+        joinBuildOutput.call,
+        joinBuildOutput.value,
+    );
+
+    return {
+        joinQueryResult,
+        joinBuildOutput,
+        txOutput,
+    };
+}
+
+export function assertUnbalancedJoin(
+    chainId: ChainId,
+    poolStateInput: PoolStateInput,
+    joinInput: UnbalancedJoinInput,
+    joinResult: JoinResult,
+    slippage: Slippage,
+) {
+    const { txOutput, joinQueryResult, joinBuildOutput } = joinResult;
+
+    // Get an amount for each pool token defaulting to 0 if not provided as input (this will include BPT token if in tokenList)
+    const expectedAmountsIn = poolStateInput.tokens.map((t) => {
+        let token;
+        if (
+            joinInput.useNativeAssetAsWrappedAmountIn &&
+            t.address === NATIVE_ASSETS[chainId].wrapped
+        )
+            token = new Token(chainId, zeroAddress, t.decimals);
+        else token = new Token(chainId, t.address, t.decimals);
+        const input = joinInput.amountsIn.find(
+            (a) => a.token.address === t.address,
+        );
+        if (input === undefined) return TokenAmount.fromRawAmount(token, 0n);
+        else return TokenAmount.fromRawAmount(token, input.amount);
+    });
+
+    const expectedQueryResult: Omit<JoinQueryResult, 'bptOut' | 'bptIndex'> = {
+        // Query should use same amountsIn as input
+        amountsIn: expectedAmountsIn,
+        tokenInIndex: undefined,
+        // Should match inputs
+        poolId: poolStateInput.id,
+        poolType: poolStateInput.type,
+        fromInternalBalance: !!joinInput.fromInternalBalance,
+        joinKind: joinInput.kind,
+    };
+
+    const queryCheck = getCheck(joinQueryResult, true);
+
+    expect(queryCheck).to.deep.eq(expectedQueryResult);
+
+    // Expect some bpt amount
+    expect(joinQueryResult.bptOut.amount > 0n).to.be.true;
+
+    assertJoinBuildOutput(
+        joinInput,
+        joinQueryResult,
+        joinBuildOutput,
+        true,
+        slippage,
+    );
+
+    assertTokenDeltas(
+        poolStateInput,
+        joinInput,
+        joinQueryResult,
+        joinBuildOutput,
+        txOutput,
+    );
+}
+
+export function assertSingleToken(
+    chainId: ChainId,
+    poolStateInput: PoolStateInput,
+    joinInput: SingleAssetJoinInput,
+    joinResult: JoinResult,
+    slippage: Slippage,
+) {
+    const { txOutput, joinQueryResult, joinBuildOutput } = joinResult;
+
+    if (!joinQueryResult.tokenInIndex) throw Error('No index');
```

Comment:
Maybe change this to an assertion where `tokenInIndex` should be defined (and even >= 0 to make sure it's not -1)?
If you think it should indeed be throwing like this, I'd update it to check for `!== undefined` because the current check will throw for index `0`.

Reply:
This was actually just to narrow the type rather than testing anything but definitely need to keep the 0 index in mind so changed it to check for undefined. Thanks.